src/utils/car.js

```diff
@@ -33,20 +37,49 @@ const hashes = {
 export async function validateBody (body) {
   const carBlockIterator = await CarBlockIterator.fromIterable(body)
   for await (const { cid, bytes } of carBlockIterator) {
-    if (!codecs[cid.code]) {
-      throw new Error(`Unexpected codec: 0x${cid.code.toString(16)}`)
-    }
-    if (!hashes[cid.multihash.code]) {
-      throw new Error(`Unexpected multihash code: 0x${cid.multihash.code.toString(16)}`)
-    }
-
-    // Verify step 2: if we hash the bytes, do we get the same digest as reported by the CID?
-    // Note that this step is sufficient if you just want to safely verify the CAR's reported CIDs
-    const hash = await hashes[cid.multihash.code].digest(bytes)
-    if (toHex(hash.digest) !== toHex(cid.multihash.digest)) {
-      throw new Error('Hash mismatch')
-    }
+    await verifyBlock(cid, bytes)
   }
 
   return true
 }
+
+/**
+ * Verifies a block
+ *
+ * @param {CID} cid
+ * @param {Uint8Array} bytes
+ */
+export async function verifyBlock (cid, bytes) {
+  // Verify step 1: is this a CID we know how to deal with?
+  if (!codecs[cid.code]) {
+    throw new VerificationError(`Unexpected codec: 0x${cid.code.toString(16)}`)
+  }
+  if (!hashes[cid.multihash.code]) {
+    throw new VerificationError(`Unexpected multihash code: 0x${cid.multihash.code.toString(16)}`)
+  }
+
+  // Verify step 2: if we hash the bytes, do we get the same digest as
+  // reported by the CID? Note that this step is sufficient if you just
+  // want to safely verify the CAR's reported CIDs
+  const hash = await hashes[cid.multihash.code].digest(bytes)
+  if (toHex(hash.digest) !== toHex(cid.multihash.digest)) {
+    throw new VerificationError(
+          `Mismatch: digest of bytes (${toHex(hash)}) does not match digest in CID (${toHex(cid.multihash.digest)})`)
+  }
+}
+
+/**
+ * Verifies and extracts the raw content from a CAR stream.
+ *
+ * @param {string} cidPath
+ * @param {ReadableStream|AsyncIterable} carStream
+ */
+export async function * extractVerifiedContent (cidPath, carStream) {
+  const getter = await CarBlockGetter.fromStream(carStream)
+
+  for await (const child of recursive(cidPath, getter)) {
```

Comment:
Here's another annoying thing that you're going to need to test for - you can't assume all your data will be unixfs, and I'm not sure if `recursive()` here is going to be happy when you feed it non-unixfs data.

When building out your test suite for this, you need to include some other DAG forms that are not unixfs. Some set of dag-cbor blocks that you can path through and verify that it works transparently of whether you're dealing with unixfs or not. The annoying subtlety is that we've first-classed unixfs in this protocol, but allowed for non-unixfs DAGs. So your pathing mechanisms need to change depending on what you're faced with. `recursive` here should be fine with the unixfs special-cased pathing rules, but when you have non-unixfs data, you need to follow standard ipld pathing through the nodes of each block. Some details on the differences are covered [here](https://github.com/ipld/ipld/blob/master/specs/codecs/dag-pb/spec.md#alternative-legacy-pathing).

Thankfully, in most cases, non-unixfs pathing is just a naive ordered walk through the JavaScript forms to find the links to follow, like the various operations in https://github.com/multiformats/js-multiformats/blob/master/src/block.js that take a `path` (specifically `get` might do the trick).

Reply:
Is this the tyranny of unixfs that I've been hearing about 😆 

Is there an existing lib I should be using, besides the js-multiformats you linked? Maybe helia? All I need to do here is walk the dag and export the data, right? I feel like that generic logic should exist somewhere in the js ipfs ecosystem.

Do cid file paths (`<cid>/cat.png`) exist in other DAG types besides unixfs?

Reply:
How do I check if a CID is unixfs or not?

Reply:
`recursive` seems to work fine with non-unixfs dag-cbor CARs. I tested it with this CAR: https://github.com/ipfs/gateway-conformance/blob/main/fixtures/path_gateway_dag/dag-cbor-traversal.car. That's a non-unixfs DAG.. right?

In fact I don't think the dag-cbor logic even implements unixfs logic, it looks like a basic traversal implementation. There's no logic that handles directories or files. https://github.com/ipfs/js-ipfs-unixfs/blob/master/packages/ipfs-unixfs-exporter/src/resolvers/dag-cbor.ts